lib/fileset.nix

```diff
@@ -0,0 +1,993 @@
+  _differenceTree = path: lhs: rhs:
+    # Branch 1
+    if isString rhs || lhs == null then
+      null
+    # Branch 2
+    else if rhs == null then
+      lhs
+    # Branch 3
+    else
+      mapAttrs (name: lhsValue:
+        _differenceTree (append path name) lhsValue rhs.${name}
+      ) (_directoryEntries path lhs);
+
+  # Whether two <tree>'s are equal
+  # Type: Path -> <tree> -> <tree> -> <tree>
+  #
+  # | lhs\rhs |   null  |   attrs |   dir  |   str   |
+  # | ------- | ------- | ------- | ------ | ------- |
+  # | null    | 1 true  | 1 rec   | 1 rec  | 1 false |
+  # | attrs   | 2 rec   | 3 rec   | 3 rec  |   -     |
+  # | dir     | 2 rec   | 3 rec   | 4 true |   -     |
+  # | str     | 2 false |   -     |   -    | 4 true  |
+  _equalsTree = path: lhs: rhs:
+    # Branch 1
+    if lhs == null then
+      _isEmptyTree path rhs
+    # Branch 2
+    else if rhs == null then
+      _isEmptyTree path lhs
+    # Branch 3
+    else if isAttrs lhs || isAttrs rhs then
+      let
+        lhs' = _directoryEntries path lhs;
+        rhs' = _directoryEntries path rhs;
+      in
+      all (name:
+        _equalsTree (append path name) lhs'.${name} rhs'.${name}
+      ) (attrNames lhs')
+    # Branch 4
+    else
+      true;
+
+  # Whether a tree is empty, containing no files
+  # Type: Path -> <tree> -> Bool
+  _isEmptyTree = path: tree:
+    if isAttrs tree || tree == "directory" then
+      let
+        entries = _directoryEntries path tree;
+      in
+      all (name: _isEmptyTree (append path name) entries.${name}) (attrNames entries)
+    else
+      tree == null;
+
+  # Simplifies a tree, optionally expanding all "directory"'s into complete listings
+  # Type: Bool -> Path -> <tree> -> <tree>
+  _simplifyTree = expand: base: tree:
+    let
+      recurse = focusPath: tree:
+        if tree == "directory" && expand || isAttrs tree then
+          let
+            expanded = _directoryEntries focusPath tree;
+            transformedSubtrees = mapAttrs (name: recurse (append focusPath name)) expanded;
+            values = attrValues transformedSubtrees;
+          in
+          if all (value: value == "emptyDir") values then
+            "emptyDir"
+          else if all (value: isNull value || value == "emptyDir") values then
+            null
+          else if !expand && all (value: isString value || value == "emptyDir") values then
+            "directory"
+          else
+            mapAttrs (name: value: if value == "emptyDir" then null else value) transformedSubtrees
+        else
+          tree;
+      result = recurse base tree;
+    in
+    if result == "emptyDir" then
+      null
+    else
+      result;
+
+  _prettyTreeSuffix = tree:
+    if isAttrs tree then
+      ""
+    else if tree == "directory" then
+      " (recursive directory)"
+    else
+      " (${tree})";
+
+  # Pretty-print all files included in the file set.
+  # Type: (b -> String -> b) -> b -> Path -> FileSet -> b
+  _prettyFoldl' = f: start: base: tree:
```

Comment:
We discussed this. This is such a clever abuse of the language that I'm not even sure if tracing should be part of the interface. It is convenient though. Maybe we could do a follow-up PR to add tracing, but first factor out the specifics of tracing from the tree fold.